R/mergedatasetsbycase.R

```diff
@@ -2603,20 +2603,33 @@ mergeSrc <- function(n.data.set.cases, data.set.names, existing.mergesrc = NULL)
     mergesrc <- rep(seq_len(n.data.sets), n.data.set.cases)
     attr(mergesrc, "labels") <- structure(seq_len(n.data.sets),
                                         .Names = data.set.names)
-    # Shift new mergesrc values and append to existing
-    # mergesrc variable.
-    if (!is.null(existing.mergesrc)) {
-        existing.labels <- attr(existing.mergesrc, "labels")
-        mergesrc[!is.na(existing.mergesrc)] <- NA
+    existing.labels <- attr(existing.mergesrc, "labels")
+
+    # existing.mergesrc is the mergesrc variable created my merging the input
+    # data sets. It exists when an input data set contains a mergesrc because
+    # it was created from a previous data set merge.
+    # We incorporate existing.mergesrc into the mergesrc of this merge by
+    # reusing the labels from existing.mergesrc when they are not missing,
+    # and adding labels based on the input data set names when they are missing.
+    # Note that the "labels" attribute is a named integer vector whose values
+    # correspond to those in mergesrc, and whose names are the actual "labels".
+    if (!is.null(existing.mergesrc) && !is.null(existing.labels)) {
```

Comment:
`&& !is.null(existing.labels)` is the extra check and the fix. The rest of the changes are refactoring.